src/contribute/portability.md

```diff
@@ -0,0 +1,164 @@
+   for how to build container images.
+
+In discussion:
+
+ * [Cross
+   language](https://s.apache.org/beam-mixed-language-pipelines). Options
+   and tradeoffs for how to handle various kinds of
+   multi-language/multi-SDK pipelines.
+
+## Development
+
+The portability framework is a substantial effort that touches every
+Beam component. Other than the sheer magnitude, a major challenge is
+engineering an interop layer that is not significantly compromising
+performance due to the additional serialization overhead of a
+language-neutral protocol.
+
+### Roadmap
+
+The proposed project phases are roughly as follows and are not
+strictly sequential, not least because various components will likely
+move at different speeds. Additionally, there have been (and continues
+to be) supporting refactorings that are not always tracked as part of
+the portability effort. Work already done is not tracked here either.
+
+ * **P1 [MVP]**: Implement the fundamental plumbing for portable SDKs
+   and runners for batch and streaming, including containers and the
+   ULR
+   [[BEAM-2899](https://issues.apache.org/jira/browse/BEAM-2899)]. Each
+   SDK and runner should use the portability framework at least to the
+   extent that wordcount
+   [[BEAM-2896](https://issues.apache.org/jira/browse/BEAM-2896)] and
+   windowed wordcount
+   [[BEAM-2941](https://issues.apache.org/jira/browse/BEAM-2941)] run
+   portably.
+    
+ * **P2 [Feature complete]**: Design and implement portability support
+   for remaining execution-side features, so that any pipeline from
+   any SDK can run portably on any runner. These features include side
+   inputs
+   [[BEAM-2863](https://issues.apache.org/jira/browse/BEAM-2863)], User
+   timers
+   [[BEAM-2925](https://issues.apache.org/jira/browse/BEAM-2925)],
+   Splittable DoFn
+   [[BEAM-2896](https://issues.apache.org/jira/browse/BEAM-2896)] and
+   more.  Each SDK and runner should use the portability framework at
+   least to the extent that the mobile gaming examples
+   [[BEAM-2940](https://issues.apache.org/jira/browse/BEAM-2940)] run
+   portably.
+    
+ * **P3 [Performance]**: Measure and tune performance of portable
+   pipelines using benchmarks such as Nexmark. Features such as
+   progress reporting
+   [[BEAM-2940](https://issues.apache.org/jira/browse/BEAM-2940)],
+   combiner lifting
+   [[BEAM-2937](https://issues.apache.org/jira/browse/BEAM-2937)] and
+   fusion are expected to be needed.
+
+ * **P4 [Cross language]**: Design and implement cross-language
+   pipeline support, including how the ecosystem of shared transforms
+   should work.
+
+### Issues
+
+The portability effort touches every component, so the "portability"
+label is used to identify all portability-related issues. Pure
+design or proto definitions should use the "beam-model" component. A
+common pattern for new portability features is that the overall
+feature is in "beam-model" with subtasks for each SDK and runner in
+their respective
```

Comment:
not sure if this sentence is missing the end or the JIRA down below is the end. JIRA is on a separate line so a bit off either way. that JIRA link does not work for me -- it says the filter does not exist or is private, even when logged in.